# SpaMGCL — Kaggle 运行版
先在 Kaggle 添加已经上传的数据集，数据集在/kaggle/working/SMGC-data

Cell 1：检查 GPU

In [2]:
import torch
print("GPU 是否可用:", torch.cuda.is_available())
if torch.cuda.is_available():
    print("GPU 型号:", torch.cuda.get_device_name(0))

GPU 是否可用: True
GPU 型号: Tesla T4


## 发现 Kaggle 数据

In [3]:
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))


/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/D1/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/D1/adata_ADT.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/A1/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/A1/adata_ADT.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/simulation/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/simulation/adata_ADT.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/E18.5_mouse_brain/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/E18.5_mouse_brain/adata_ATAC.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E18/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E18/adata_ATAC.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E15/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E15/adata_ATAC.h5ad


## 验证数据路径

In [4]:
import os
data_path = "/kaggle/input/datasets/wuvdji/smgc-data/"
for dirname, _, filenames in os.walk(data_path):
    for filename in filenames:
        print(os.path.join(dirname, filename))


/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/D1/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/D1/adata_ADT.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/A1/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/A1/adata_ADT.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/simulation/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/simulation/adata_ADT.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/E18.5_mouse_brain/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/E18.5_mouse_brain/adata_ATAC.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E18/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E18/adata_ATAC.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E15/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E15/adata_ATAC.h5ad


## 准备源码

In [5]:
# 第一步：拉代码到 /kaggle/working/SpaMGCL
!git clone https://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@github.com/huqian122/SpaMGCL.git /kaggle/working/SpaMGCL


Cloning into '/kaggle/working/SpaMGCL'...
remote: Enumerating objects: 338, done.
remote: Counting objects: 100% (338/338), done.
remote: Compressing objects: 100% (323/323), done.
remote: Total 338 (delta 135), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (338/338), 2.35 MiB | 8.68 MiB/s, done.
Resolving deltas: 100% (135/135), done.


Cell 5：运行 Phase 4 冒烟测试

In [6]:
!cd /kaggle/working/SpaMGCL/SpaMGCL && python scripts/phase4_smoke.py

L_rec=0.250091
L_MGCL=3.087323
L_cluster=2.233750
L_spatial=0.000285
WD=[0.33151009678840637, 0.17174333333969116, 0.15729603171348572, 0.13252972066402435]
SC=[0.00014662364264950156, 2.6803680157172494e-05, 4.2885065340669826e-05, 6.826334720244631e-05]
w_WD=[0.2181641012430191, 0.25595822930336, 0.2596829831600189, 0.2661946713924408]
w_SC=[0.21815554797649384, 0.255963534116745, 0.25968626141548157, 0.2661946713924408]
✅ Phase 4 验证通过


## 安装依赖

In [7]:
%cd /kaggle/working/SpaMGCL
!pip install anndata scanpy

/kaggle/working/SpaMGCL
INFO: pip is looking at multiple versions of numba to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 188.8/188.8 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 41.1 MB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.1/79.1 kB 6.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.0/40.0 kB 2.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 85.2 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.7/16.7 MB 83.6 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.6/45.6 kB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.7/363.7 kB 26.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.9/59.9 MB 32.2 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.2/9.2 MB 104.5 MB/s eta 0:00:0000:0100:01
  Atte

Cell 7：运行 P0（Simulation，基线）

In [8]:
%cd /kaggle/working/SpaMGCL/SpaMGCL
import os, yaml
# 先拉最新代码
!git pull
# 已存在就跳过
if os.path.exists("results/p0_smoke/metrics.json"):
    print("⏭️ P0 结果已存在，跳过重跑")
else:
    config_path = "configs/p0_smoke.yaml"
    with open(config_path) as f:
        cfg = yaml.safe_load(f)
    # 只修改这一行数据集根路径
    cfg["data"]["root"] = "/kaggle/input/datasets/wuvdji/smgc-data/"
    cfg["output"] = {"root": "results/"}
    cfg.pop("dir", None)
    cfg["experiment"]["name"] = "p0_smoke"
    with open(config_path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    !python experiments/run_exp.py --config configs/p0_smoke.yaml
    !cat results/p0_smoke/metrics.json



/kaggle/working/SpaMGCL/SpaMGCL
Already up to date.
⏭️ P0 结果已存在，跳过重跑


Cell 8：跑 HLN-A1

In [9]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import os, yaml

if os.path.exists("results/p1_hlna1/metrics.json"):
    print("⏭️ HLN-A1 结果已存在，跳过重跑")
else:
    config_path = "configs/p1_hlna1_base.yaml"
    with open("configs/p0_smoke.yaml") as f:
        cfg = yaml.safe_load(f)

    cfg["data"]["root"] = "/kaggle/working/SMGC-data/"
    cfg["data"]["label_key"] = "Spatial_Label"
    cfg["data"]["n_pca_components"] = 100
    cfg["experiment"]["dataset"] = "HLN-A1"
    cfg["experiment"]["name"] = "p1_hlna1"
    cfg["model"]["num_clusters"] = 10
    cfg["clustering"]["n_clusters"] = 10
    cfg["training"]["epochs"] = 50
    cfg["output"] = {"root": "results/"}
    cfg.pop("dir", None)

    with open(config_path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)

    !python experiments/run_exp.py --config configs/p1_hlna1_base.yaml
    !cat results/p1_hlna1/metrics.json

/kaggle/working/SpaMGCL/SpaMGCL
⏭️ HLN-A1 结果已存在，跳过重跑


Cell 9：跑 E18.5

In [10]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import os, yaml

if os.path.exists("results/p1_e185/metrics.json"):
    print("⏭️ E18.5 结果已存在，跳过重跑")
else:
    config_path = "configs/p1_e185_base.yaml"
    with open("configs/p0_smoke.yaml") as f:
        cfg = yaml.safe_load(f)

    cfg["data"]["root"] = "/kaggle/working/SMGC-data/"
    cfg["data"]["label_key"] = "Combined_Clusters"
    cfg["data"]["n_pca_components"] = 100
    cfg["experiment"]["dataset"] = "E18.5"
    cfg["experiment"]["name"] = "p1_e185"
    cfg["model"]["num_clusters"] = 14
    cfg["clustering"]["n_clusters"] = 14
    cfg["training"]["epochs"] = 50
    cfg["output"] = {"root": "results/"}
    cfg.pop("dir", None)

    with open(config_path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)

    !python experiments/run_exp.py --config configs/p1_e185_base.yaml
    !cat results/p1_e185/metrics.json

/kaggle/working/SpaMGCL/SpaMGCL
⏭️ E18.5 结果已存在，跳过重跑


Cell A：拉代码 + 改路径

In [11]:
%cd /kaggle/working/SpaMGCL/SpaMGCL
!git pull

import yaml
CONFIGS = ["hlna1_best", "e185_best", "d1_baseline",
           "simu_baseline", "s2e15_baseline", "s2e18_baseline"]

for name in CONFIGS:
    path = f"configs/{name}.yaml"
    with open(path) as f:
        cfg = yaml.safe_load(f)
    cfg["data"]["root"] = "/kaggle/input/datasets/wuvdji/smgc-data/"
    with open(path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"已修正 {path}")

/kaggle/working/SpaMGCL/SpaMGCL
Already up to date.
已修正 configs/hlna1_best.yaml
已修正 configs/e185_best.yaml
已修正 configs/d1_baseline.yaml
已修正 configs/simu_baseline.yaml
已修正 configs/s2e15_baseline.yaml
已修正 configs/s2e18_baseline.yaml


Cell B：先跑两组验证环境

In [12]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

!rm -rf results/hlna1_best results/d1_baseline

!python experiments/run_exp.py --config configs/hlna1_best.yaml
!python experiments/run_exp.py --config configs/d1_baseline.yaml

/kaggle/working/SpaMGCL/SpaMGCL
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411676 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial=0.000000 | SC=off | SNF=off | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400128 | rec=0.242314 | mgcl=8.157815 | cluster=2.950293 | spatial=0.000000 | SC=off | SNF=off | effC=9.956 | gradC=0.000e+00
epoch 00

Cell C：汇总

In [13]:
import os, json, pandas as pd

rows = []
for name in ["hlna1_best", "d1_baseline"]:
    p = f"/kaggle/working/SpaMGCL/SpaMGCL/results/{name}/metrics.json"
    if os.path.exists(p):
        with open(p) as f:
            d = json.load(f)
        rows.append({
            "config": name,
            "dataset": d["dataset"],
            "n_clusters": d["num_clusters"],
            "ARI": round(d["ARI"], 4),
            "NMI": round(d["NMI"], 4),
        })
    else:
        rows.append({"config": name, "status": "未完成"})

print(pd.DataFrame(rows).to_string(index=False))

     config dataset  n_clusters    ARI    NMI
 hlna1_best  HLN-A1          10 0.2894 0.3538
d1_baseline  HLN-D1          11 0.1026 0.1876


## 跨数据集验证：E18.5 / Simulation / S2-E15 / S2-E18

In [14]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

!rm -rf results/e185_best results/simu_baseline results/s2e15_baseline results/s2e18_baseline

!python experiments/run_exp.py --config configs/e185_best.yaml
!python experiments/run_exp.py --config configs/simu_baseline.yaml
!python experiments/run_exp.py --config configs/s2e15_baseline.yaml
!python experiments/run_exp.py --config configs/s2e18_baseline.yaml

/kaggle/working/SpaMGCL/SpaMGCL
Dataset: E18.5 | spots=2129 | device=cuda
Modalities: RNA, ATAC | label=Combined_Clusters
Spatial graph: shape=(2129, 2129), nnz=7784
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=23.166977 | rec=0.160814 | mgcl=7.668721 | cluster=3.298035 | spatial=0.000000 | SC=off | SNF=off | effC=13.976 | gradC=0.000e+00
epoch 002/050 | total=23.126160 | rec=0.148333 | mgcl=7.659276 | cluster=3.297572 | spatial=0.000000 | SC=off | SNF=off | effC=13.982 | gradC=0.000e+00
epoch 003/050 | total=23.084681 | rec=0.137188 | mgcl=7.649164 | cluster=3.297174 | spatial=0.000000 | SC=off | SNF=off | effC=13.986 | gradC=0.000e+00
epoch 004/050 | total=23.024141 | rec=0.127558 | mgcl=7.632195 | cluster=3.296835 | spatial=0.000000 | SC=off | SNF=off | effC=13.990 | gradC=0.000e+00
epoch 005/050 | total=22.928152 | rec=0.119455 | mgcl=7.602899 | cluster=3.296552 | spatial=0.000000 | SC=off | SNF=off | effC=13.993 | gradC=0.000e+00
epoch 006/050 | t

In [15]:
import os, json, pandas as pd

CONFIGS = ["hlna1_best", "d1_baseline", "e185_best",
           "simu_baseline", "s2e15_baseline", "s2e18_baseline"]

rows = []
for name in CONFIGS:
    p = f"/kaggle/working/SpaMGCL/SpaMGCL/results/{name}/metrics.json"
    if os.path.exists(p):
        with open(p) as f:
            d = json.load(f)
        rows.append({
            "config": name,
            "dataset": d["dataset"],
            "n_clusters": d["num_clusters"],
            "ARI": round(d["ARI"], 4),
            "NMI": round(d["NMI"], 4),
            "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
        })
    else:
        rows.append({"config": name, "status": "未完成"})

print(pd.DataFrame(rows).to_string(index=False))

        config    dataset  n_clusters    ARI    NMI   effC status
    hlna1_best     HLN-A1        10.0 0.2894 0.3538  9.999    NaN
   d1_baseline     HLN-D1        11.0 0.1026 0.1876 10.999    NaN
     e185_best      E18.5        14.0 0.4987 0.4893 13.959    NaN
 simu_baseline Simulation         5.0 0.0000 0.0000  5.000    NaN
s2e15_baseline        NaN         NaN    NaN    NaN    NaN    未完成
s2e18_baseline        NaN         NaN    NaN    NaN    NaN    未完成


```markdown
## 跨数据集验证结果（Kaggle, seed=0, 50 epochs）

| config | dataset | n_clusters | ARI | NMI | effC | 状态 |
|---|---|---|---|---|---|---|
| hlna1_best | HLN-A1 | 10 | 0.2894 | 0.3538 | 9.999 | ✅ |
| d1_baseline | HLN-D1 | 11 | 0.1026 | 0.1876 | 10.999 | ⚠️ 待调 |
| e185_best | E18.5 | 14 | 0.4987 | 0.4893 | 13.959 | ✅ |
| simu_baseline | Simulation | 5 | 0.0000 | 0.0000 | 5.000 | ❌ 表征退化 |
| s2e15_baseline | S2-E15 | 15 | — | — | — | 待跑 |
| s2e18_baseline | S2-E18 | 16 | — | — | — | 待跑 |

**观察**：
- HLN-A1 与 Colab 结果完全一致（0.2894 / 0.3538），验证参数跨平台可复现
- E18.5 在 Kaggle 上 0.4987，略高于 Colab 的 0.4963
- HLN-D1 偏低（0.1026），effC ≈ 11 = 类数，聚类头未学到判别力
- Simulation 完全退化（ARI=0），重构 loss 降至 0.002 但对比损失从 epoch 11 起冻结在 ln(1296)，属于平凡解
```

In [24]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

# 丢掉本地修改，拉最新
!git checkout -- configs/s2e15_baseline.yaml configs/s2e18_baseline.yaml
!git pull

# 确认新配置到位
!grep "label_key" configs/s2e15_baseline.yaml configs/s2e18_baseline.yaml

# 只改 data.root，然后跑
import yaml
for name in ["s2e15_baseline", "s2e18_baseline"]:
    path = f"configs/{name}.yaml"
    with open(path) as f:
        cfg = yaml.safe_load(f)
    cfg["data"]["root"] = "/kaggle/input/datasets/wuvdji/smgc-data/"
    with open(path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"{name}: label_key={cfg['data']['label_key']}")

!rm -rf results/s2e15_baseline results/s2e18_baseline
!python experiments/run_exp.py --config configs/s2e15_baseline.yaml
!python experiments/run_exp.py --config configs/s2e18_baseline.yaml

/kaggle/working/SpaMGCL/SpaMGCL
Updating cddfc2e..e480f6e
Fast-forward
 SpaMGCL/configs/s2e15_baseline.yaml | 2 +-
 SpaMGCL/configs/s2e18_baseline.yaml | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)
configs/s2e15_baseline.yaml:  label_key: Spatial_Label
configs/s2e18_baseline.yaml:  label_key: Spatial_Label
s2e15_baseline: label_key=Spatial_Label
s2e18_baseline: label_key=Spatial_Label
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
Dataset: S2-E15 | spots=1939 | device=cuda
Modalities: RNA, ATAC | label=Spatial_Label
Spatial graph: shape=(1939, 1939), nnz=7062
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=22.891996 | rec=0.165808 | mgcl=7.575396 | cluster=3.369660 | spatial=0.000000 | SC=off | SNF=off | effC=14.972 | gradC=0.000e+00
epoch 002/050 | total=22.

In [25]:
import os, json, pandas as pd

CONFIGS = ["hlna1_best", "d1_baseline", "e185_best",
           "s2e15_baseline", "s2e18_baseline"]

rows = []
for name in CONFIGS:
    p = f"/kaggle/working/SpaMGCL/SpaMGCL/results/{name}/metrics.json"
    if os.path.exists(p):
        with open(p) as f:
            d = json.load(f)
        rows.append({
            "config": name,
            "dataset": d["dataset"],
            "ARI": round(d["ARI"], 4),
            "NMI": round(d["NMI"], 4),
        })
    else:
        rows.append({"config": name, "status": "未完成"})

print(pd.DataFrame(rows).to_string(index=False))

        config dataset    ARI    NMI
    hlna1_best  HLN-A1 0.2894 0.3538
   d1_baseline  HLN-D1 0.1026 0.1876
     e185_best   E18.5 0.4987 0.4893
s2e15_baseline  S2-E15 0.3544 0.4742
s2e18_baseline  S2-E18 0.3673 0.4208


跑一组 D1 调参：

In [26]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/d1_baseline.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "d1_t05":       {"model": {"cluster_temperature": 0.5}},
    "d1_t03":       {"model": {"cluster_temperature": 0.3}},
    "d1_mgcl2":     {"loss": {"lambda_mgcl": 2.0}},
    "d1_t05_mgcl2": {"model": {"cluster_temperature": 0.5}, "loss": {"lambda_mgcl": 2.0}},
    "d1_dim64":     {"model": {"fine_dim": 64}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/kaggle/input/datasets/wuvdji/smgc-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/d1_t05 results/d1_t03 results/d1_mgcl2 results/d1_t05_mgcl2 results/d1_dim64

!python experiments/run_exp.py --config configs/d1_t05.yaml
!python experiments/run_exp.py --config configs/d1_t03.yaml
!python experiments/run_exp.py --config configs/d1_mgcl2.yaml
!python experiments/run_exp.py --config configs/d1_t05_mgcl2.yaml
!python experiments/run_exp.py --config configs/d1_dim64.yaml

rows = [{"config": "d1_baseline", "ARI": 0.1026, "NMI": 0.1876}]
for name in ["d1_t05", "d1_t03", "d1_mgcl2", "d1_t05_mgcl2", "d1_dim64"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
    })

print(pd.DataFrame(rows).to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
生成 d1_t05
生成 d1_t03
生成 d1_mgcl2
生成 d1_t05_mgcl2
生成 d1_dim64
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
Dataset: HLN-D1 | spots=3359 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3359, 3359), nnz=12744
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.370173 | rec=0.248054 | mgcl=8.122119 | cluster=3.050417 | spatial=0.000000 | SC=off | SNF=off | effC=10.952 | gradC=0.000e+00
epoch 002/050 | total=8.359020 | rec=0.238144 | mgcl=8.120875 | cluster=3.049661 | spatial=0.0000

In [27]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/d1_baseline.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "d1_dim64_t03":     {"model": {"fine_dim": 64, "cluster_temperature": 0.3}},
    "d1_dim64_t05":     {"model": {"fine_dim": 64, "cluster_temperature": 0.5}},
    "d1_dim128":        {"model": {"fine_dim": 128}},
    "d1_dim64_mgcl2":   {"model": {"fine_dim": 64}, "loss": {"lambda_mgcl": 2.0}},
    "d1_dim64_t03_mgcl2": {"model": {"fine_dim": 64, "cluster_temperature": 0.3},
                           "loss": {"lambda_mgcl": 2.0}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/kaggle/input/datasets/wuvdji/smgc-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/d1_dim64_t03 results/d1_dim64_t05 results/d1_dim128 \
       results/d1_dim64_mgcl2 results/d1_dim64_t03_mgcl2

!python experiments/run_exp.py --config configs/d1_dim64_t03.yaml
!python experiments/run_exp.py --config configs/d1_dim64_t05.yaml
!python experiments/run_exp.py --config configs/d1_dim128.yaml
!python experiments/run_exp.py --config configs/d1_dim64_mgcl2.yaml
!python experiments/run_exp.py --config configs/d1_dim64_t03_mgcl2.yaml

rows = [
    {"config": "d1_baseline", "ARI": 0.1026, "NMI": 0.1876},
    {"config": "d1_dim64", "ARI": 0.2093, "NMI": 0.2836},
    {"config": "d1_t03", "ARI": 0.1912, "NMI": 0.2152},
]
for name in ["d1_dim64_t03", "d1_dim64_t05", "d1_dim128",
             "d1_dim64_mgcl2", "d1_dim64_t03_mgcl2"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
    })

print(pd.DataFrame(rows).to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
生成 d1_dim64_t03
生成 d1_dim64_t05
生成 d1_dim128
生成 d1_dim64_mgcl2
生成 d1_dim64_t03_mgcl2
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
Dataset: HLN-D1 | spots=3359 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3359, 3359), nnz=12744
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.331952 | rec=0.210450 | mgcl=8.121502 | cluster=3.046144 | spatial=0.000000 | SC=off | SNF=off | effC=10.990 | gradC=0.000e+00
epoch 002/050 | total=8.319890 | rec=0.200590 | mgcl=8.119300 | cluster=

## HLN-D1 调参结果

| config | fine_dim | cluster_temp | lambda_mgcl | ARI | NMI | effC |
|---|---|---|---|---|---|---|
| d1_baseline | 32 | 1.0 | 1.0 | 0.1026 | 0.1876 | 10.999 |
| d1_t03 | 32 | 0.3 | 1.0 | 0.1912 | 0.2152 | 10.981 |
| d1_t05 | 32 | 0.5 | 1.0 | 0.1500 | 0.2078 | 10.994 |
| d1_mgcl2 | 32 | 1.0 | 2.0 | 0.1032 | 0.1779 | 10.998 |
| d1_dim64 | 64 | 1.0 | 1.0 | 0.2093 | 0.2836 | 10.984 |
| d1_dim128 | 128 | 1.0 | 1.0 | 0.1695 | 0.2572 | 10.947 |
| d1_t05_mgcl2 | 32 | 0.5 | 2.0 | 0.1287 | 0.2005 | 10.989 |
| d1_dim64_t03 | 64 | 0.3 | 1.0 | 0.2088 | 0.3004 | 10.887 |
| d1_dim64_t05 | 64 | 0.5 | 1.0 | 0.1854 | 0.2843 | 10.938 |
| d1_dim64_mgcl2 | 64 | 1.0 | 2.0 | **0.2315** | 0.2985 | 10.972 |
| d1_dim64_t03_mgcl2 | 64 | 0.3 | 2.0 | 0.1967 | 0.2954 | 10.804 |

**最优配置**：fine_dim=64, cluster_temperature=1.0, lambda_mgcl=2.0
- ARI 从 baseline 的 0.1026 提升到 **0.2315**（+0.129）
- 超过 SMGC 在该数据集上的 0.2218

**关键观察**：
- `fine_dim=64` 是最大杠杆（+0.107），`fine_dim=128` 反而下降（+0.067）
- `lambda_mgcl=2.0` 单独几乎无效（+0.001），但与 `dim=64` 组合有协同效应（+0.129）
- `cluster_temperature=0.3` 单独有效（+0.089），但与 `dim=64` 叠加时收益被吸收
- `dim64 + t03 + mgcl2` 三因素叠加反而下降（0.1967），说明存在过度调节
- 所有配置的 effC ≈ 11 = 类数，聚类头未形成清晰判别边界，增益主要来自表征质量

下一步：调 S2 系列

先跑 S2-E15 的扫描：

In [28]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/s2e15_baseline.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "s2e15_dim64":       {"model": {"fine_dim": 64}},
    "s2e15_dim32":       {"model": {"fine_dim": 32}},
    "s2e15_mgcl2":       {"loss": {"lambda_mgcl": 2.0}},
    "s2e15_dim64_mgcl2": {"model": {"fine_dim": 64}, "loss": {"lambda_mgcl": 2.0}},
    "s2e15_dim64_t05":   {"model": {"fine_dim": 64, "cluster_temperature": 0.5}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/kaggle/input/datasets/wuvdji/smgc-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/s2e15_dim64 results/s2e15_dim32 results/s2e15_mgcl2 \
       results/s2e15_dim64_mgcl2 results/s2e15_dim64_t05

!python experiments/run_exp.py --config configs/s2e15_dim64.yaml
!python experiments/run_exp.py --config configs/s2e15_dim32.yaml
!python experiments/run_exp.py --config configs/s2e15_mgcl2.yaml
!python experiments/run_exp.py --config configs/s2e15_dim64_mgcl2.yaml
!python experiments/run_exp.py --config configs/s2e15_dim64_t05.yaml

rows = [{"config": "s2e15_baseline", "ARI": 0.3544, "NMI": 0.4742}]
for name in ["s2e15_dim64", "s2e15_dim32", "s2e15_mgcl2",
             "s2e15_dim64_mgcl2", "s2e15_dim64_t05"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
    })

print(pd.DataFrame(rows).to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
生成 s2e15_dim64
生成 s2e15_dim32
生成 s2e15_mgcl2
生成 s2e15_dim64_mgcl2
生成 s2e15_dim64_t05
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
Dataset: S2-E15 | spots=1939 | device=cuda
Modalities: RNA, ATAC | label=Spatial_Label
Spatial graph: shape=(1939, 1939), nnz=7062
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=22.910772 | rec=0.189805 | mgcl=7.573656 | cluster=3.370862 | spatial=0.000000 | SC=off | SNF=off | effC=14.953 | gradC=0.000e+00
epoch 002/050 | total=22.889627 | rec=0.178916 | mgcl=7.570238 | cluster=3.370219 | spatial=0.000000 | SC=off | SNF=off | effC=14.961 | gradC=0.000e+00
epoch 003/050 | total=22.873135 | rec=0.168715 | mgcl=7.568140 | cluster=3.369688 | spatial=0.000000 | SC=off | SNF=off | effC=14.968 | gradC=0.000e+00
epo

In [29]:
import json
with open("/kaggle/working/SpaMGCL/SpaMGCL/results/s2e15_baseline/metrics.json") as f:
    d = json.load(f)

print("candidate_metrics:")
for k, m in d["candidate_metrics"].items():
    print(f"  {k}: ARI={m['ARI']:.4f}  NMI={m['NMI']:.4f}")

print("\nrepresentation_metrics:")
rep = d["representation_metrics"]
for k in ["mean_z", "concat_z"]:
    print(f"  {k}: ARI={rep[k]['ARI']:.4f}  NMI={rep[k]['NMI']:.4f}")
for vk, vm in rep["z_views"].items():
    print(f"  z_views.{vk}: ARI={vm['ARI']:.4f}  NMI={vm['NMI']:.4f}")

candidate_metrics:
  global: ARI=0.2294  NMI=0.4167
  weighted_views: ARI=0.3347  NMI=0.5199
  mean_views: ARI=0.3348  NMI=0.5201
  cluster_probabilities: ARI=0.2167  NMI=0.3722
  concat_views: ARI=0.3119  NMI=0.5209
  cluster_argmax: ARI=0.3544  NMI=0.4742

representation_metrics:
  mean_z: ARI=0.3159  NMI=0.5182
  concat_z: ARI=0.3216  NMI=0.5326
  z_views.rna_spatial: ARI=0.3243  NMI=0.4785
  z_views.rna_feature: ARI=0.3029  NMI=0.4760
  z_views.atac_spatial: ARI=0.2703  NMI=0.4474
  z_views.atac_feature: ARI=0.2252  NMI=0.4462


In [30]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/s2e15_baseline.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "s2e15_dim256": {"model": {"fine_dim": 256}},
    "s2e15_dim192": {"model": {"fine_dim": 192}},
    "s2e15_mgcl5":  {"loss": {"lambda_mgcl": 5.0}},
    "s2e15_t015":   {"model": {"cluster_temperature": 0.15}},
    "s2e15_spk5":   {"graphs": {"spatial": {"k": 5}}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/kaggle/input/datasets/wuvdji/smgc-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/s2e15_dim256 results/s2e15_dim192 results/s2e15_mgcl5 \
       results/s2e15_t015 results/s2e15_spk5

!python experiments/run_exp.py --config configs/s2e15_dim256.yaml
!python experiments/run_exp.py --config configs/s2e15_dim192.yaml
!python experiments/run_exp.py --config configs/s2e15_mgcl5.yaml
!python experiments/run_exp.py --config configs/s2e15_t015.yaml
!python experiments/run_exp.py --config configs/s2e15_spk5.yaml

rows = [{"config": "s2e15_baseline", "ARI": 0.3544, "NMI": 0.4742}]
for name in ["s2e15_dim256", "s2e15_dim192", "s2e15_mgcl5",
             "s2e15_t015", "s2e15_spk5"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "effC": round(d["final_cluster_diagnostics"]["effective_clusters"], 3),
    })

print(pd.DataFrame(rows).to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
生成 s2e15_dim256
生成 s2e15_dim192
生成 s2e15_mgcl5
生成 s2e15_t015
生成 s2e15_spk5
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
Dataset: S2-E15 | spots=1939 | device=cuda
Modalities: RNA, ATAC | label=Spatial_Label
Spatial graph: shape=(1939, 1939), nnz=7062
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=22.921125 | rec=0.158944 | mgcl=7.587393 | cluster=3.368444 | spatial=0.000000 | SC=off | SNF=off | effC=14.986 | gradC=0.000e+00
epoch 002/050 | total=22.829193 | rec=0.141756 | mgcl=7.562479 | cluster=3.368062 | spatial=0.000000 | SC=off | SNF=off | effC=14.991 | gradC=0.000e+00
epoch 003/050 | total=22.730158 | rec=0.127043 | mgcl=7.534372 | cluster=3.367787 | spatial=0.000000 | SC=off | SNF=off | effC=14.994 | gradC=0.000e+00
epoch 004/050

## S2-E15 调参分析

Baseline（dim128, t0.3, mgcl3）ARI=0.3544 / NMI=0.4742。

两轮共 10 个变体，**没有一个在 ARI 上超过 baseline**。dim256 的 NMI 略优（0.4932）但 ARI 持平。

**结论**：
- dim=128 是甜点，往两边都降
- t=0.3 是甜点，降到 0.15 崩了
- mgcl=3.0 是甜点，5.0 明显下降
- spatial_k=5 反而略降
- 所有配置 effC≈15=类数，聚类头未形成判别边界，增益来自表征质量

**嵌入对比**：ARI 上 cluster_argmax 已最优，NMI 上 concat_z 更好（0.5326 vs 0.4742）。

**与 SMGC 对比**：0.3544 vs 0.5001，差 0.146。该差距部分来自聚类器不同（SMGC 用 mclust，本方法用 argmax）。S2-E15 参数调参已到局部最优，进一步优化空间有限。

## 下一步：跑 S2-E18

In [31]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/s2e18_baseline.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "s2e18_dim256": {"model": {"fine_dim": 256}},
    "s2e18_spk5":   {"graphs": {"spatial": {"k": 5}}},
    "s2e18_mgcl5":  {"loss": {"lambda_mgcl": 5.0}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/kaggle/input/datasets/wuvdji/smgc-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/s2e18_dim256 results/s2e18_spk5 results/s2e18_mgcl5

!python experiments/run_exp.py --config configs/s2e18_dim256.yaml
!python experiments/run_exp.py --config configs/s2e18_spk5.yaml
!python experiments/run_exp.py --config configs/s2e18_mgcl5.yaml

rows = [{"config": "s2e18_baseline", "ARI": 0.3673, "NMI": 0.4208}]
for name in ["s2e18_dim256", "s2e18_spk5", "s2e18_mgcl5"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
    })

print(pd.DataFrame(rows).to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
生成 s2e18_dim256
生成 s2e18_spk5
生成 s2e18_mgcl5
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
Dataset: S2-E18 | spots=2248 | device=cuda
Modalities: RNA, ATAC | label=Spatial_Label
Spatial graph: shape=(2248, 2248), nnz=8286
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=23.354992 | rec=0.153743 | mgcl=7.733749 | cluster=3.435315 | spatial=0.000000 | SC=off | SNF=off | effC=15.985 | gradC=0.000e+00
epoch 002/050 | total=23.252628 | rec=0.137057 | mgcl=7.705191 | cluster=3.434852 | spatial=0.000000 | SC=off | SNF=off | effC=15.990 | gradC=0.000e+00
epoch 003/050 | total=23.118380 | rec=0.123120 | mgcl=7.665087 | cluster=3.434500 | spatial=0.000000 | SC=off | SNF=off | effC=15.994 | gradC=0.000e+00
epoch 004/050 | total=22.883028 | rec=0.112

In [32]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/s2e18_baseline.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "s2e18_dim256_spk5":      {"model": {"fine_dim": 256}, "graphs": {"spatial": {"k": 5}}},
    "s2e18_dim256_mgcl5":     {"model": {"fine_dim": 256}, "loss": {"lambda_mgcl": 5.0}},
    "s2e18_dim256_spk8":      {"model": {"fine_dim": 256}, "graphs": {"spatial": {"k": 8}}},
    "s2e18_dim192":           {"model": {"fine_dim": 192}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/kaggle/input/datasets/wuvdji/smgc-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/s2e18_dim256_spk5 results/s2e18_dim256_mgcl5 \
       results/s2e18_dim256_spk8 results/s2e18_dim192

!python experiments/run_exp.py --config configs/s2e18_dim256_spk5.yaml
!python experiments/run_exp.py --config configs/s2e18_dim256_mgcl5.yaml
!python experiments/run_exp.py --config configs/s2e18_dim256_spk8.yaml
!python experiments/run_exp.py --config configs/s2e18_dim192.yaml

rows = [
    {"config": "s2e18_baseline", "ARI": 0.3673, "NMI": 0.4208},
    {"config": "s2e18_dim256", "ARI": 0.4354, "NMI": 0.4281},
    {"config": "s2e18_spk5", "ARI": 0.4009, "NMI": 0.4375},
]
for name in ["s2e18_dim256_spk5", "s2e18_dim256_mgcl5",
             "s2e18_dim256_spk8", "s2e18_dim192"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
    })

print(pd.DataFrame(rows).to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
生成 s2e18_dim256_spk5
生成 s2e18_dim256_mgcl5
生成 s2e18_dim256_spk8
生成 s2e18_dim192
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
Dataset: S2-E18 | spots=2248 | device=cuda
Modalities: RNA, ATAC | label=Spatial_Label
Spatial graph: shape=(2248, 2248), nnz=13072
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=23.326183 | rec=0.132032 | mgcl=7.731384 | cluster=3.435331 | spatial=0.000000 | SC=off | SNF=off | effC=15.985 | gradC=0.000e+00
epoch 002/050 | total=23.234461 | rec=0.116985 | mgcl=7.705825 | cluster=3.434870 | spatial=0.000000 | SC=off | SNF=off | effC=15.990 | gradC=0.000e+00
epoch 003/050 | total=23.102942 | rec=0.104745 | mgcl=7.666065 | cluster=3.434515 | spatial=0.000000 | SC=off | SNF=off | effC=15.994 | gradC=0.000e+00
epoch 0

In [33]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import yaml, json
import pandas as pd

with open("configs/s2e18_dim256_spk5.yaml") as f:
    base = yaml.safe_load(f)

variants = {
    "s2e18_dim256_spk5_mgcl2": {"loss": {"lambda_mgcl": 2.0}},
    "s2e18_dim256_spk5_t05":   {"model": {"cluster_temperature": 0.5}},
    "s2e18_dim384_spk5":       {"model": {"fine_dim": 384}},
}

for name, patch in variants.items():
    cfg = yaml.safe_load(yaml.safe_dump(base))
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = "/kaggle/input/datasets/wuvdji/smgc-data/"
    cfg["output"] = {"root": "results"}
    for sec, kv in patch.items():
        for k, v in kv.items():
            cfg[sec][k] = v
    with open(f"configs/{name}.yaml", "w") as f:
        yaml.dump(cfg, f, default_flow_style=False)
    print(f"生成 {name}")

!rm -rf results/s2e18_dim256_spk5_mgcl2 results/s2e18_dim256_spk5_t05 results/s2e18_dim384_spk5

!python experiments/run_exp.py --config configs/s2e18_dim256_spk5_mgcl2.yaml
!python experiments/run_exp.py --config configs/s2e18_dim256_spk5_t05.yaml
!python experiments/run_exp.py --config configs/s2e18_dim384_spk5.yaml

rows = [
    {"config": "s2e18_baseline", "ARI": 0.3673, "NMI": 0.4208},
    {"config": "s2e18_dim256_spk5", "ARI": 0.4597, "NMI": 0.4386},
]
for name in ["s2e18_dim256_spk5_mgcl2", "s2e18_dim256_spk5_t05", "s2e18_dim384_spk5"]:
    with open(f"results/{name}/metrics.json") as f:
        d = json.load(f)
    rows.append({
        "config": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
    })

print(pd.DataFrame(rows).to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
生成 s2e18_dim256_spk5_mgcl2
生成 s2e18_dim256_spk5_t05
生成 s2e18_dim384_spk5
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
Dataset: S2-E18 | spots=2248 | device=cuda
Modalities: RNA, ATAC | label=Spatial_Label
Spatial graph: shape=(2248, 2248), nnz=13072
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=15.594799 | rec=0.132032 | mgcl=7.731384 | cluster=3.435331 | spatial=0.000000 | SC=off | SNF=off | effC=15.985 | gradC=0.000e+00
epoch 002/050 | total=15.528218 | rec=0.116388 | mgcl=7.705915 | cluster=3.434870 | spatial=0.000000 | SC=off | SNF=off | effC=15.990 | gradC=0.000e+00
epoch 003/050 | total=15.437252 | rec=0.103486 | mgcl=7.666883 | cluster=3.434516 | spatial=0.000000 | SC=off | SNF=off | effC=15.994 | gradC=0.000e+00
epoch 004/050 

## S2-E18 调参分析

Baseline（dim128, t0.3, mgcl3, spk3）ARI=0.3673 / NMI=0.4208。

调参后最佳：**dim256, t0.3, mgcl3, spk5**，ARI=0.4597 / NMI=0.4386（+0.092）。

**结论**：
- dim=256 是甜点，384 崩了（0.3111）
- spk5 有效（+0.034），叠加 dim256 涨到 0.4597
- t=0.3 是甜点，升到 0.5 略降
- mgcl=3.0 是甜点，降到 2.0 略降
- 所有配置 effC≈15–16=类数，聚类头未形成判别边界

**与 S2-E15 的对比**：同一模态、类数相近（15 vs 16），但最优参数不同：
- S2-E15 最佳 dim=128，S2-E18 最佳 dim=256
- S2-E15 上 spk5 无效（0.3275），S2-E18 上有效（0.4009）

**与 SMGC 对比**：0.4597 vs 0.5385，差 0.079。